csrc/lower_thread_predicate.cpp

```diff
@@ -506,6 +516,70 @@ void ThreadPredicateMap::build(Fusion* fusion) {
   for (auto expr : fusion->exprs()) {
     updateBitSet(expr);
   }
+
+  // record output tensor and its correspoinding domain merged from broadcast
+  // domain(s)
+  std::vector<std::pair<const TensorView*, IterDomain*>>
+      merged_broadcast_domains;
+  std::vector<const TensorView*> out_tv_without_broadcast_domain;
+  for (auto out_val : fusion->outputs()) {
+    auto out_tv = dynamic_cast<const TensorView*>(out_val);
+    bool tensor_with_domain_merged_from_broadcast = false;
+    auto from = out_tv->getMaybeRFactorDomain();
+    auto all_exp = DependencyCheck::getAllExprsBetween(
+        {from.begin(), from.end()},
+        {out_tv->domain()->domain().begin(), out_tv->domain()->domain().end()});
+    for (auto expr : all_exp) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto domain = merge->out();
+        auto pt = domain->getParallelType();
+        bool from_broadcast =
+            merge->outer()->isBroadcast() || merge->inner()->isBroadcast();
+        if (isParallelTypeThread(pt) && from_broadcast) {
```

Comment:
I revised the algorithm, it can capture the case you mensioned. Current alg:
(1) starting at an output tensor, if one of its leaf node's definition is a merge and parallelized by thread/block, move to (2)
(2) check all the root domains, if it is a broadcast, map it to its concretized domain  (or concretized domain in its exact map)
(3) Loop over all leaf nodes whose definition is a merge and parallelized by thread/block
---(3.1) find all root domains that are merged to this leaf domain.
---(3.2) Set stride when use extent of merged root domains to index leaf domain.
```
  // e.g. Root: [I1,B2,B3] -> Leaf: [I1*B2*B3], the merged_root_domains =
  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. For broadcast
  // root domain, use the extent of its concretized domain. 
```
---(3.3)  set write stride
```
    // write_stride_mod will generate condition: index % pair(1) < pair(2)
    std::unordered_map<ParallelType, std::pair<Val*, Val*>> write_stride_mod;
    // write_stride_less will generate condition: index < write_stride_less
    std::unordered_map<ParallelType, Val*> write_stride_less;
```
7 addtional test cases are added:
```
  // Test case where [B1,I2,I3] is merged to [B1I2I3] and paralled by blockIdx.x
  // The write pattern should be: write only the first len(I2)*len(I3) blocks,
  // where len(I) is the extent of the concretized domain.
  // write_stride_less= ( ( 1 * T0.size[1] ) * T0.size[0] )
  // condition = blockIdx.x < ( ( 1 * T0.size[1] ) * T0.size[0] )
  runTest({true, false, false, false});

  // Test case where [I1,B2,I3] is merged to [I1B2I3] and paralled by blockIdx.x
  // The write pattern should be: len(I3) blocks write, in every len(B2)*len(I3)
  // blocks, write_stride_mod= {len(B2)*len(I3), len(I3)} condition = blockIdx.x
  // % (len(B2)*len(I3)) < len(I3)
  runTest({false, true, false, false});

  // Test case where [I1,I2,B3] is merged to [I1I2B3] and paralled by blockIdx.x
  // The write pattern should be: write every len(B3) blocks,
  // where len(B) is the extent of the domain.
  // write_stride_mod= ( 1 * T1.size[2] )
  runTest({false, false, true, false});

  // Test case where [I1,B2,B3] is merged to [I1B2B3] and paralled by blockIdx.x
  // The write pattern should be: write every len(B2)*len(B3) blocks,
  // where len(B) is the extent of the concretized domain.
  // write_stride_mod= ( ( 1 * T1.size[2] ) * T1.size[1] )
  runTest({false, true, true, false});

  // Test case where [B1,I2,B3] is merged to [B1I2B3] and paralled by blockIdx.x
  // The write pattern should be: write every len(B3) blocks of the first
  // len(I2) * len(B3) blocks write_stride_less= ( ( 1 * T1.size[2] ) *
  // T0.size[0] ) write_stride_mod= ( 1 * T1.size[2] )
  runTest({true, false, true, false});

  // Test case where [B1,B2,I3] is merged to [B1B2I3] and paralled by blockIdx.x
  // The write pattern should be: write only the first len(I3) blocks
  // write_stride_less= ( 1 * T0.size[0] )
  runTest({true, true, false, false});

  // Test case where [B1,B2,B3] is merged to [B1B2B3] and paralled by blockIdx.x
  // The write pattern should be: write only the first block
  // write_stride_less= 1 (write when blockIDx.x < 1)
  runTest({true, true, true, false});
```